crates/move-stackless-bytecode/src/macro_quantifiers_analysis.rs

```diff
@@ -0,0 +1,259 @@
+                    let callee_id = mod_id.qualified(*fun_id);
+                    return (attr_id.clone(), dsts.clone(), srcs.clone(), callee_id, type_params.clone());
+                }
+            },
+            _ => {}
+        };
+
+        unreachable!("extract_fn_call_data should only be called with function call bytecode")
+    }
+
+    fn is_fn_call(&self, bc: &Bytecode) -> bool {
+        match bc {
+            Bytecode::Call(_, _, operation, _, _) => {
+                match operation {
+                    Operation::Function(_,_, _) => true,
+                    _ => false
+                }
+            },
+            _ => false,
+        }
+    }
+
+    fn is_destroy(&self, bc: &Bytecode) -> bool {
+        match bc {
+            Bytecode::Call(_, _, op, _, _) => matches!(op, Operation::Destroy),
+            _ => false,
+        }
+    }
+
+    fn is_trace_local(&self, bc: &Bytecode) -> bool {
+        match bc {
+            Bytecode::Call(_, _, op, _, _) => matches!(op, Operation::TraceLocal(_)),
+            _ => false,
+        }
+    }
+
+    fn is_searched_fn(&self, bc: &Bytecode, qid: QualifiedId<FunId>) -> bool {
+        match bc {
+            Bytecode::Call(_, _, operation, _, _) => {
+                match operation {
+                    Operation::Function(mod_id,fun_id, _) => {
+                        qid == mod_id.qualified(*fun_id)
+                    },
+                    _ => false
+                }
+            },
+            _ => false,
+        }
+    }
+
+    fn validate_function_pattern_requirements(&self, env: &GlobalEnv, targets: &FunctionTargetsHolder, qid: QualifiedId<FunId>) -> bool {
+        let func_env = env.get_function(qid);
+        let data = targets.get_data(&qid, &FunctionVariant::Baseline).unwrap();
+
+        if !no_abort_analysis::get_info(data).does_not_abort && !targets.is_abort_check_fun(&qid) {
+            env.diag(
+                Severity::Error,
+                &func_env.get_loc(),
+                "Quantififier function should not abort",
+            );
+
+            return true;
+        }
+
+        if !deterministic_analysis::get_info(data).is_deterministic {
+            env.diag(
+                Severity::Error,
+                &func_env.get_loc(),
+                "Quantififier function should be deterministic",
+            );
+
+            return true;
+        }
+
+        // probably other check in future
+        return false;
+    }
+
+    pub fn find_macro_patterns(&self, env: &GlobalEnv, targets: &FunctionTargetsHolder, start_qid: QualifiedId<FunId>, end_qid: QualifiedId<FunId>, quantifier: QuantifierType, bc: &Vec<Bytecode>) -> Vec<Bytecode> {
+        let chain_len = 5;
+        if bc.len() < chain_len {
+            return bc.to_vec();
+        }
+
+        for i in 0..bc.len() - (chain_len - 1) {
+            if 
+                self.is_searched_fn(&bc[i], start_qid) &&
```

Comment:
when encountering a `start_qid`, the rest of the pattern should match or there should be an error